pkg/controllers/inventory/endpoint_handler.go

```diff
@@ -0,0 +1,166 @@
+			c.handleEndpoint(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.handleEndpoint(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.handleEndpoint(obj)
+		},
+	})
+	if err != nil {
+		log.Error(err, "Failed to add Endpoints event handler")
+		return err
+	}
+	return nil
+}
+
+func (c *InventoryController) handleEndpoint(obj interface{}) {
+	var endpoint *v1.Endpoints
+	ok := false
+	switch obj1 := obj.(type) {
+	case *v1.Endpoints:
+		endpoint = obj1
+	case cache.DeletedFinalStateUnknown:
+		endpoint, ok = obj1.Obj.(*v1.Endpoints)
+		if !ok {
+			err := fmt.Errorf("obj is not valid *v1.Endpoints")
+			log.Error(err, "DeletedFinalStateUnknown Obj is not *v1.Endpoints")
+			return
+		}
+	}
+	log.V(1).Info("Inventory processing Endpoint", "Endpoint", endpoint.Name, "Namespace", endpoint.Namespace)
+	service := &v1.Service{}
+	err := c.Client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      endpoint.Name,
+			Namespace: endpoint.Namespace,
+		},
+		service,
+	)
+	if err != nil {
+		log.Error(err, "Failed to get Service", "Name", endpoint.Name, "Namespace", endpoint.Namespace)
+		return
+	}
+	c.triggerPod(service)
+	c.triggerService(service)
+}
+
+func (c *InventoryController) triggerService(service *v1.Service) {
+	key, _ := keyFunc(service)
+	log.V(1).Info("Adding Service key to inventory object queue", "Service key", key)
+	c.inventoryObjectQueue.Add(inventory.InventoryKey{InventoryType: inventory.ContainerApplication, ExternalId: string(service.UID), Key: key})
+}
+
+func (c *InventoryController) triggerPod(service *v1.Service) {
+	// Get endpoint's corresponding pods and enqueue them
+	podIds, _ := c.service.GetPodIDsFromEndpoint(service)
+	for _, podID := range podIds {
+		ctx := context.TODO()
+		pod, err := getPodByUID(ctx, c.Client, types.UID(podID), service.Namespace)
+		if err != nil {
+			log.Error(err, "Failed to get Pod by UID", "PodID", podID, "Namespace", service.Namespace)
+			continue
+		}
+
+		serviceUIDs, err := getServicesUIDByPodUID(ctx, c.Client, pod.UID, pod.Namespace)
+		if err != nil {
+			log.Error(err, "Failed to get services UIDs by pod UID", "Pod UID", pod.UID, "Pod Namespace", pod.Namespace)
+			continue
+		}
+
+		containerApplicationInstance := containerinventory.ContainerApplicationInstance{
+			DisplayName:             pod.Name,
+			ContainerApplicationIds: serviceUIDs,
+			ExternalId:              podID,
+		}
+		c.service.UpdatePendingAdd(podID, &containerApplicationInstance)
+
+		podKey, keyErr := keyFunc(pod)
+		if keyErr != nil {
+			log.Error(keyErr, "Failed to get key for Pod", "Pod", pod.Name, "Namespace", pod.Namespace)
+			continue
+		}
+		log.V(1).Info("Adding Endpoint ContainerApplicationInstance key to inventory object queue", "ContainerApplicationInstance ExternalId", podID, "ContainerApplicationInstance Key", podKey)
+		c.inventoryObjectQueue.Add(inventory.InventoryKey{InventoryType: inventory.ContainerApplicationInstance, ExternalId: podID, Key: podKey})
+	}
```

Comment:
it seems you still need to update the pod which not referred by the service (referred by the service before)

Reply:
fixed